arbitrum/handler_p2p.go

```diff
@@ -0,0 +1,565 @@
+		log.Warn("arbitrum_p2p: suspend while syncing", "head", h.syncedBlockNum)
+	}
+	return h.beaconBackFiller.Suspend()
+}
+
+func (h *filler) Resume() {
+	defer h.beaconBackFiller.Resume()
+	head, err := h.downloader.SkeletonHead()
+	if err != nil || head == nil {
+		log.Error("arbitrum_p2p: error from SkeletonHead", "err", err)
+		return
+	}
+	if !head.Number.IsUint64() {
+		log.Error("arbitrum_p2p: syncedBlockNum bad number", "num", head.Number)
+		return
+	}
+	h.headersLock.Lock()
+	if h.confirmed.Number.Cmp(head.Number) < 0 {
+		// confirmed only moves forward and is used as head for sync.. somerthing bad already happened
+		log.Error("arbitrum_p2p: skeleton head ahead of confirmed", "skeleton", head.Number, "confirmed", h.confirmed.Number)
+	}
+	h.syncedBlockNum = head.Number.Uint64()
+	h.syncedCond.Broadcast()
+	h.headersLock.Unlock()
+	log.Trace("arbitrum_p2p: resume", "skeletonhead", h.syncedBlockNum)
+}
+
+func (h *filler) SetMode(mode downloader.SyncMode) {
+	h.beaconBackFiller.SetMode(mode)
+}
+
+type arbHandler protocolHandler
+
+func (h *arbHandler) PeerInfo(id enode.ID) interface{} {
+	return nil
+}
+
+func (h *arbHandler) HandleLastConfirmed(peer *arb.Peer, confirmed *types.Header, node uint64) {
+	protoHandler := (*protocolHandler)(h)
+	validated := false
+	valid := false
+	current, _ := protoHandler.getHeaders()
+	if current != nil {
+		if confirmed.Number.Cmp(current.Number) == 0 {
+			validated = true
+			valid = current.Hash() == confirmed.Hash()
+		}
+	}
+	if !validated {
+		var err error
+		valid, err = h.helper.ValidateConfirmed(confirmed, node)
+		if err != nil {
+			log.Error("error in validate confirmed", "id", peer.ID(), "err", err)
+			return
+		}
+	}
+	if !valid {
+		protoHandler.peerDrop(peer.ID())
+		return
+	}
+	hPeer := protoHandler.getPeer(peer.ID())
+	if hPeer == nil {
+		log.Warn("hPeer not found on HandleLastConfirmed")
+		return
+	}
+	peer.RequestCheckpoint(nil)
+	protoHandler.advanceConfirmed(confirmed)
+}
+
+func (h *arbHandler) HandleCheckpoint(peer *arb.Peer, checkpoint *types.Header, supported bool) {
+	protoHandler := (*protocolHandler)(h)
+	log.Error("got checkpoint", "from", peer.ID(), "checkpoint", checkpoint, "supported", supported)
+	if !supported {
+		return
+	}
+	if !h.syncing.Load() {
+		return
+	}
+	if !checkpoint.Number.IsUint64() {
+		log.Warn("got bad header from peer - number not uint64", "peer", peer.ID())
+		protoHandler.peerDrop(peer.ID())
+		return
+	}
+	number := checkpoint.Number.Uint64()
+	log.Info("handler_p2p: handle checkpoint - before", "peer", peer.ID())
+	protoHandler.waitBlockSync(number)
+	log.Info("handler_p2p: handle checkpoint - after", "peer", peer.ID())
+	if !h.syncing.Load() {
+		return
+	}
+	canonical := rawdb.ReadCanonicalHash(h.db, number)
+	if canonical == (common.Hash{}) {
+		skeleton := rawdb.ReadSkeletonHeader(h.db, number)
+		if skeleton == nil {
+			log.Error("arbitrum handler_p2p: canonical not found", "number", number, "peer", peer.ID())
```

Comment:
nice catch!
we don't need to panic, just return (effectively: ignore the received checkpoint message)

Reply:
should we also drop the peer then?

I am still not sure about how we should handle this corner case, because if we can't read skeleton header from the database here, then either the database is corrupted or `waitBlockSync` didn't work.
In `waitBlockSync` (called few lines before) we are waiting until `filler.Resume` sets `h.syncedBlockNum` to number greater or equal then the checkpoint block number. `filler.Resume` reads the number from header from`h.downloader.SkeletonHead()` and it internally reads the header from the database. So it seems impossible to first get the header from db and for the second time not, unless something is broken.

Does it make sense?
And if so, should we log a critical error here (or panic)?

Reply:
I'm not sure how we could fail this test. I don't have a good scenario that doesn't require some bug somewhere.
Say we do have a race bug in waitForBlockSync - in that case, a return is the right approach. Maybe if we get another checkpoint from this or another peer it will work.

Generally if you get an unexpected error you should try to contain it. Just make sure you don't get illegal values elsewhere. Panic would be the opposite of containing the error.

No reason to assume anything wrong with this peer so a drop is unnecessary either. (Later we should have logic trying to handle peers DOSing etc)